### ISSUE TITLE
Polygon split at Dateline Code

### ISSUE THREAD

**Issue by sindizzy** [DotSpatial/DotSpatial#1407]
I have some code that I wrote for the NTS group that takes a list of coordinates that represent a polygon that crosses the dateline and creates a multi-part shape. Its not as easy as it sounds. Anyway, looking over the DS codebase I'm a little confused as to where  should add this routine. Should it be an extension?

**Comment by jany-tenaj** [comment]
For what do you need this routine?

**Comment by sindizzy** [comment]
well exactly as my description. imagine a list of coordinates that represent a polygon and this polygon crosses the dateline. If you just convert the coordinates to a polygon it does not display properly on a map with WGS84 projection. my routine creates a multi-part polygon split at the dateline such that the GIS is able to display without artifacts.

Discussion here: https://github.com/NetTopologySuite/NetTopologySuite/discussions/496

**Comment by MatthiasSchilder** [comment]
My first feel when I read the title of this issue was that this feature belongs somewhere in NTS. Because it is a little bit specific, I assume they don't want it in their 'core' geometry code, but maybe it could reside in some extension classes / extension methods?

**Comment by sindizzy** [comment]
That's all a good idea...except the NTS code in DS is a_very_old version. My coworkers and I use DS as our main GIS framework and want that kind of functionality available. So what are the options here?

**Comment by MatthiasSchilder** [comment]
DotSpatial.NTSExtension then maybe? It already contains some extension classes for geometry types from NTS. Maybe you can put your code in a PolygonExt. 

But, in my opinion, the far better solution would be to migrate DotSpatial to a current version of NTS. I'm aware that the version DotSpatial uses is at least 4 years or so old. We also use DS as our main framework, probably for quite some time in the future, so it is justified to invest some time in updating dependencies. A colleague of mine and me plan to look into the NTS issue in the next few months.

**Comment by sindizzy** [comment]
That sounds great.

**Comment by jany-tenaj** [comment]
My first thought also was that this belongs somewhere in NTS. But getting them to add things that are not in JTS might not be easy.

Matthias you don't have to look to hard. I have done the switch to NTS with my own code and it seems to work. But I got rid of the shape classes as well. Not sure what people would think of that ;)

If I remember the last time I tried to update DS.NTS the main problem was that we would have had to update the .Net Framework version we are using and Mogikanin didn't want to do that. So I stopped my efforts because this would have wrecked the build process.

@sindizzy  do you have to do this thing to your polygons when you use the dateline crossing map? I thought this was added to be able to show polygons that go over the dateline.

I'm not fond of the idea of putting these dateline functions all over the place. Maybe it would be better to make a Dateline project and put everything belonging to the dateline crossing in there?

**Comment by MatthiasSchilder** [comment]
@jany-tenaj Would you mind making your changes available as a branch or fork or something?

NTS has TargetFramework netstandard2.0. This is theoretically consumable from .NET Framework 4.6.1, but not recommended. It is safely consumable from 4.7.2 and above. We switched our production fork of DotSpatial to Framework 4.8 a while ago and did not note any problems since then. DotSpatials current TargetFramework, 4.5.2, looses support in April '22, so I think is is feasible to switch to a more recent framework. For Mogikanins sake, the last version with Framework 4.5.2 could be tagged or so.

**Comment by jany-tenaj** [comment]
@MatthiasSchilder Let me see what I can do about making the changes available. I switched to .net 4.7.2 because of NTS. 

The problem with updating DS would be that we'd have to update the continues integration build. This can only be done by Mogikanin.
Otherwise someone would have to recreate this mechanism so we don't loose the nuget uploads. I just don't have the time to do that.

**Comment by MatthiasSchilder** [comment]
Thank you!

I used this afternoon to play a little bit with GitHub Actions. I got building a single DS-assembly, packaging it and pushing it to nuget.org running fairly easy. Still, we'd probably need Mogikanins help when we want to publish it under the established package id. Do you think it's worth the time looking into this?

**Comment by jany-tenaj** [comment]
Yes I think it's worth it. I don't care whether we use the established package id or something else. As long as people can use the nuget packages everything should be alright.

**Comment by sindizzy** [comment]
@jany-tenaj I seem to remember there was an issue with the Dateline map as well. Let me give it another run. I do remember that any geoprocessing also gets broken if you use the basic coordinates-to-polygon routine.

From my experience with GIS, a polygon (or polyline for that matter) that crosses the dateline becomes rather troublesome once you want to display it and process it along with other shapes.

The one thing that truly always seems to work is to break it into a multi-part shape. Similar to how most land masses are split at the +/-180 meridian. Now the shapefile format seems to also have issues with different parts in a single feature sharing the same boundary. I think the geodb solves that but not 100% sure. What we often do is just create a multi-feature to represent said polygon.

I work in the aviation industry so we have routes flying across the dateline and airspaces that cross the dateline. It's always a challenge so I devised this algorithm that takes some assumptions and attempts to split a polygon. I'm fairly sure there are some scenarios where it will not work but I think for most it should work.

**Comment by sindizzy** [comment]
Also I'm not advocating putting this algorithm all over the place. It's a main routine with 2 or 3 support functions. Currently I have it in the Analysis category but can easily change it to the recommend NTS Extension category.

Since we often use this type of conversion I wanted to have it as part of the core DS. Putting it in one project is not going to help as we have many projects that require it. A single code source would help to not fragment it.

For me, I see it as any other helper routines. Points to polygon, polyline to point, points to raster, etc. This in essence is a specialized points to polygon that attempts to account for crossing of the dateline. 

geometries = PolygonizerThing(list of coords)

The dev then decides what to do with the geometry or geometries. 

I can see if NTS wants it in their codebase. If not then it would make sense to add it to DS. If they do, then we can delete it from DS once the NTS adds it to the framework.

I'm open to ideas.

**Comment by MatthiasSchilder** [comment]
> But I think GitHub Actions doesn't check whether the code is compileable? So that means we'd rather have to check the Continious integration build via Appveyor or something similar. I'd say knowing the code is compileable is quite important.

My current prototype runs msbuild on push, so you get immediate feedback. I bet there is a way to do some static code analysis and/or unit tests execution attached to a git hook. I'll take a look.

**Comment by jany-tenaj** [comment]
@sindizzy what I meant about another project was the fact that we have DatelineCrossingMap in Controls and would add your function to NtsExtensions. So we have two things belonging to the same subject in two projects. 

The reason I don't like this is the following: I was looking at zooming and found some functions in Controls and some in the scaling plugin and again some in my own code. So when you want to change something on it you have to look all over the place.

Basically I think dateline crossing is a subject that is important for some people but not for most. So having it in its own project would give those people that need it all the functions we have for dateline crossing in the same place. And all others don't have to bother with it.

**Comment by sindizzy** [comment]
I think I understand your position @jany-tenaj. Let me talk this over with my coworker as he used that dateline map more than I did. He stopped using it for some reasons I can't remember.

**Comment by sindizzy** [comment]
Ok I have some further info. A discussion with a coworker was focused on Extent.IsCrossDateline. Through his testing he found that it didnt work as desired or maybe how we thought it should work. It is set to TRUE for features that cross the prime meridian as well. I'm going to take a look at what the code does but that's another bit that's connected to the dateline.

**Comment by sindizzy** [comment]
So it looks like the IsCrossDateline routine is to be used only in coordination with the DatelineMap control.

As it is written, any Extent will expose IsCrossDateline. Thus, we thought it would work to tell us when **any** polygon crossed the dateline since it has an extent as well.

Also, maybe this would be a good location to put my code?

![image](https://user-images.githubusercontent.com/5298460/144336285-944aaec7-cd84-4860-9256-55b2a5425f19.png)

**Comment by jany-tenaj** [comment]
So IsCrossDateline can't be used for anything other than the DatelineMap? Then it shouldn't be an extension method. And no I don't think this is a good place. Methods for polygons aren't controls ;) But I know where you're coming from, as all Dateline classes are in Controls. That's why I said I'd like to put them into their own project.

**Comment by jany-tenaj** [comment]
Can Normalized be used for other things than DatelineMap? If not I think we should get rid of those extension methods and include them in DatelineCrossingMap.

And then I'd say lets wait whether Felix includes your functions in NTS before we decide where to put them in DS.

**Comment by sindizzy** [comment]
Let me take a look but I want to say I use normalize in code and it works like it should.

**Comment by sindizzy** [comment]
Ok so it looks like the dateline Normalised is only used in the dateline control. The comments say the left edge is normalised into the range -180 to +180.

The functions that I have used are from DotSpatial.Positioning.Azimuth.Normalize and ".Angle.Normalize.

The 2 latter functions are the same code and normalize the value between 0 and 360.

I think I also had wrote up an issue with setting a standard with US English vs UK English routine/variable names. Issue #1179 in the summer of 2018.

**Comment by sindizzy** [comment]
So I had a couple of minutes to show you exactly our issue. Say we want to show the world-wide established airspaces (these are very easily found online at any aviation db site).

First thing we do is convert the vertices (or coordinates) to polygons. We create one shapefile that employs the standard process. Pretty much just pGeom = new Polygon(new LinearRing(coords)). The second shapefile that is created goes through my special routine to split polygons that cross the dateline.

Now we take a **regular map control** and add

shapefile with standard processing. As you can see, those shapes that cross the dateline want to tear across the map instead of wrapping around. This is the expected rendering in any GIS that I have used since they are just connecting the dots.
![RegMap_StdProc](https://user-images.githubusercontent.com/5298460/144732473-27fa29a6-f084-4295-9749-b26824f49c67.png)

shapefile with special processing. In this case we need to tell the GIS that certain polygons cross the dateline. The best way to do that, in my experience, is to split the polygon into multiple parts.
![RegMap_SpcProc](https://user-images.githubusercontent.com/5298460/144732480-efeeebb7-7ea9-4a06-91cc-b0e907adac39.png)

Now we take a **Dateline map control** and add

shapefile with standard processing. again, the GIS draws across the map.
![DatelineMap_StdProc](https://user-images.githubusercontent.com/5298460/144732484-3e970732-55fc-431e-a79a-8c9e4c87a913.png)

shapefile with special processing
![DatelineMap_SpcProc](https://user-images.githubusercontent.com/5298460/144732489-e77f9cb0-bab8-4bf2-9d36-4aa7e6caf02a.png)

The only drawback that I can see is that the split line is visible in the Dateline map. 

I have also tried to shift the coordinates to the 0 to 360 range and I seem to remember that works but would need to go back and test. 

Another thing is that geoprocessing also didn't function correctly for certain cases. Say I have a point and want to see if it is contained within an airspace polygon. I seem to remember issues as the polygon is seen as closing towards the prime meridian and not the dateline.

**Comment by jany-tenaj** [comment]
@sindizzy maybe you could get rid of the thick line by making the dateline map show only up to 179.9° instead of 180°? Not sure what this would do with the calculations, but I think they should work anyway.

**Comment by jany-tenaj** [comment]
@MatthiasSchilder have a look at the [NtsUpdate branch](https://github.com/DotSpatial/DotSpatial/tree/NtsUpdate). I switched to the current NTS nuget release.

**Comment by sindizzy** [comment]
@jany-tenaj not sure what you mean. The dateline map shows all longitudes. How can I only go up to 179.9? Wouldn't this defeat the purpose of showing polygons traversing the dateline?

**Comment by jany-tenaj** [comment]
My thought was to make the dateline map slightly smaller, so that the dateline itsself wouldn't be seen and therefore the borderlines of your polygons wouldn't be seen. If you don't have polygons that have borders not comming from cutting them on the dateline that might get rid of the unwanted lines splitting your polygons.

**Comment by sindizzy** [comment]
Ahh I see what you mean. So the dateline map control itself would be from say -179.9999 to +179.9999. That may work but I think it's too much effort. As it is, feature classes from our provider already have the split as I mentioned and it doesn't seem to cause an issue. They come from the biggest commercial GIS company so I think they realized a long time ago that they needed to split any polygon along the dateline. If they didn't then every single map would show artifacts when the land masses were added.

Examples:

![image](https://user-images.githubusercontent.com/5298460/144767350-231ce034-a75a-467f-a9b7-0ae04a96a8a0.png)
![image](https://user-images.githubusercontent.com/5298460/144767468-962af96d-8dea-4202-94ce-5edbf117664f.png)

**Comment by sindizzy** [comment]
Btw, great discussion. The dateline has been a pain in my rump for a very long time. We render maps and process data on a worldwide basis so maybe that's why it's more important to me than some other folks. Routes cross the dateline, say from Dallas to Hong Kong. Polygons cross the dateline, say the airspaces that I have shown. So it is a real GIS challenge.

One other thing I have thought about is why the need for a special map control to show a "dateline" map. What I do with other GIS frameworks is to just use the Plate Carree projection, which essentially is just the WGS84 "projection" with a modifiable central meridian. Here is a snapshot with central meridian set to 180. This uses another GIS but DotSpatial has a definition for this projection so I am wondering what the difference is between this projection and the effects of the Dateline map control.

![image](https://user-images.githubusercontent.com/5298460/144767617-3a6d8155-ec12-413f-97ad-cf05f6a9433a.png)

**Comment by jany-tenaj** [comment]
No idea who came up with the dateline map or why. If nobody had mentioned it in some issue I might no even know that it's there ;) I sure as hell won't need it.

But I'd guess it exists because everything seems to be based on Wgs 84 projection and there the normal map doesn't work on the dateline?

**Comment by sindizzy** [comment]
ok following the suggestions here I have made a PR to the NTS repo. Not sure if they will make it part of a lib or leave it as a sample workflow. Anyway, take a look when you get a chance: https://github.com/NetTopologySuite/NetTopologySuite/pull/580

**Comment by jany-tenaj** [comment]
so can we close this?
